# <font color=red>Solutions for</font> Exercises for Lab 03! <font color=blue>Version 0.8 Beta </font>

## Functions used for Tables:
Just like in Lab 02 Exercises, the bulk of this first part is to provide a reference for the functions you'll use. The tables here are taken from the datascience module, so all these table functions can only work if you import datascience (see the first line of code after this).

Remember that in a given table, lists make up the rows and arrays make up columns. Additionally, tables have separate indexes for rows and columns. An index of 0 could refer to the first column or first row (the header row is not given an index) depending on the function. However, if you're typing `existingTable[index]`, it will go for columns, not rows.

### Functions
|Purpose|Function|
|--|--|
|Creating a New Table (from arrays)|`newTable = Table().with_columns(columnHeader1, array1, columnHeader2, array2, [etc, etc])` (it's much recommended to make a table using arrays as columns rather than using lists as rows.)|
|Creating a New Table (from .csv file)|`newTable = Table.read_table('fileName')`|
|Revealing the full table|`existingTable.show(numberOfRows)` use empty parentheses if you want all rows|
|Restricting a Table to certain parameters|`existingTable.where(columnHeader, are.condition(parameter))` (check below for conditions)|
|Selecting  row in Table format|`existingTable.take(indexOfRow1, indexOfRow2, [etc, etc])` (you can use functions to create arrays to put here for the row indices)|
|Selecting a column in Table format|`existingTable.select('columnHeader1' OR indexOfColumn1, 'columnHeader2' OR indexOfColumn2, [etc, etc])` OR `existingTable[indexOfColumn]`
|Selecting a column in array format|`existingTable.column('columnHeader1' OR indexOfColumn1, 'columnHeader2' OR indexOfColumn2, [etc, etc])`|
|Adding one row to a Table|`existingTable = existingTable.with_row(newRowAsList1, newRowAsList2, [etc, etc])`|
|Adding one column to a Table|`existingTable = existingTable.with_column(columnHeader1, array1, columnHeader2, array2, [etc, etc])`|
|Adding multiple rows to a Table|`existingTable = existingTable.with_rows(newRowAsList1, newRowAsList2, [etc, etc])`|
|Adding multiple columns to a Table|`existingTable = existingTable.with_columns(columnHeader1, array1, columnHeader2, array2, [etc, etc])`|
|Sorting the rows|`existingTable.sort(columnHeader, descending = True/False)` (descending is automatically set to false; put it to True if you want highest values first)|
|Getting the number of rows/columns|`existingTable.num_rows` OR `existingTable.num_columns`|
|Getting rid of certain columns|`existingTable.drop(unwantedColumnHeader1, unwantedColumnHeader2, [etc, etc])`|

### Conditions
|Purpose|Function (in np.where)|
|--|--|
|where target values are equal to the parameter|`existingTable.where(columnHeader, are.equal_to(parameter))` OR `existingTable.where(columnHeader, parameter)`|
|where target values are NOT equal to the parameter|`existingTable.where(columnHeader, are.not_equal_to(parameter))`|
|where target values are less than the parameter|`existingTable.where(columnHeader, are.below(parameter))`|
|where target values are more than the parameter|`existingTable.where(columnHeader, are.above(parameter))`|
|where target values are between two parameters|`existingTable.where(columnHeader, are.between(parameter1, parameter2))` (this includes values equal to parameter1, but NOT those equal to parameter2)|
|where target values contain the parameter|`existingTable.where(columnHeader, are.containing(parameter))`|


One key detail is that you need to keep track of indexes. Taking the index of a table once gives you a column, and taking the index twice gives you a cell. For example, `existingTable[5][2]` goes to the table's 6th column (remember, indexes start counting at 0, not 1) and gets the cell of that column that is in the 3rd row.

Before we can even hope to work with tables, we must import the relevant libraries. These relevant libraries are the datascience module, which were will almost always import without its name.

In [2]:
from datascience import *
import numpy as np # this library is also used in these Exercises.

## Exercise 1: 
A central part of data manipulation is, well, the manipulation of data (well done Einstein). This primarily constitutes the use of FOM (Functions, Operations, and/or Methods) to show another aspect of data or to express existing data in a different format.

Why do we want to manipulate data?
* To remake it in a format that is more easily usable by the computer itself (Remember, your capabilities in analyzing data are determined at least partially by the capabilities of Jupyter itself!)
* To conduct statistical tests and other advanced actions (spoilers for Lab 05 and onwards)

The principal way to manipulate data is to extract it from the Table setting. While it may seem that Tables are the all-important focus of this class, don't fall for that illusion. **Tables are only a convenient way to organize and understand data. By themselves, they are incapable of actual analysis.** The true leaders here are the lists and arrays. I'll leave it to you to find the exact method that converts tables to arrays in the above table.

### 1.1: Title Length
Create the imdb Table from the file named 'imdb.csv', and extract the Title column as an array. 

In [3]:
imdb = Table.read_table('imdb.csv')
titles = imdb.column('Title')

Use the relevant command, and the correct format, to get an array that contains the number of characters in each title. For example, the movie title 'León' would beget a value of 4 (it has 4 characters). Review the Exercises 02 if you do not know.

In [4]:
num_chars = [len(x) for x in titles]

Append this column to the imdb table. Give it any column header you want. :']

In [5]:
imdb = imdb.with_column('Title Length', num_chars)
imdb

Votes,Rating,Title,Year,Decade,Title Length
88355,8.4,M,1931,1930,1
132823,8.3,Singin' in the Rain,1952,1950,19
74178,8.3,All About Eve,1950,1950,13
635139,8.6,Léon,1994,1990,4
145514,8.2,The Elephant Man,1980,1980,16
425461,8.3,Full Metal Jacket,1987,1980,17
441174,8.1,Gone Girl,2014,2010,9
850601,8.3,Batman Begins,2005,2000,13
37664,8.2,Judgment at Nuremberg,1961,1960,21
46987,8,Relatos salvajes,2014,2010,16


What are the five movies with the longest names?

In [6]:
imdb.sort('Title Length', descending=True)

Votes,Rating,Title,Year,Decade,Title Length
309141,8.5,Dr. Strangelove or: How I Learned to Stop Worrying and L ...,1964,1960,68
718645,8,Pirates of the Caribbean: The Curse of the Black Pearl,2003,2000,54
1099087,8.8,The Lord of the Rings: The Fellowship of the Ring,2001,2000,49
700283,8.7,Star Wars: Episode V - The Empire Strikes Back,1980,1980,46
1074146,8.9,The Lord of the Rings: The Return of the King,2003,2000,45
466111,8,Harry Potter and the Deathly Hallows: Part 2,2011,2010,44
556811,8.4,Star Wars: Episode VI - Return of the Jedi,1983,1980,42
372863,8.2,"Lock, Stock and Two Smoking Barrels (1998)",1998,1990,42
447875,8.9,"Il buono, il brutto, il cattivo (1966)",1966,1960,38
967389,8.7,The Lord of the Rings: The Two Towers,2002,2000,37


Dr. Strangelove, PotC Black Pearl, LOTR Fellowship of the Ring, Star Wars V, and LOTR Return of the King

### 1.2: Multistep process
Add a column to the imdb table that shows how many a's that movie title has. For example, the movie title 'All About Eve' would have a value of 2.

I'll walk you through this one, since it does a good job of showing how multistep coding procedures work in regards to data manipulation.

In [7]:
imdb

Votes,Rating,Title,Year,Decade,Title Length
88355,8.4,M,1931,1930,1
132823,8.3,Singin' in the Rain,1952,1950,19
74178,8.3,All About Eve,1950,1950,13
635139,8.6,Léon,1994,1990,4
145514,8.2,The Elephant Man,1980,1980,16
425461,8.3,Full Metal Jacket,1987,1980,17
441174,8.1,Gone Girl,2014,2010,9
850601,8.3,Batman Begins,2005,2000,13
37664,8.2,Judgment at Nuremberg,1961,1960,21
46987,8,Relatos salvajes,2014,2010,16


Step 1: You may have noticed that the letter A sometimes appears as uppercase (A) and other times as lowercase (a). Python treats these two as totally distinct characters (like apples and oranges). Extract the 'Title' column as an array again (or use the same variable if you already defined it in Exercise 1.1) and use the method `String.lower()` to convert all characters to lowercase.

In [8]:
lowercase_titles = [x.lower() for x in titles]

Step 2: Now comes the heart of the problem, that is, how to make Python count all the a's in a title. There are many ways to go about this, and you are welcome to create your own solution here. However, one of the simplest I know revolves around the `String.split(splitBoundary)` method, which was introduced in the Exercises for Lab 02. Refer back there (particularly to the 2.3 and 2.4 exercise problems) or to your own toolbox (that you *should* create) for how to use that method.

Note that, in the example below, for each instance of A, the method effectively cleaves the string at that location. Thus, you can expect that a word with one A begets two lists when put in the .split method, two A's begets three lists, and so forth.

Your lightbulb may already be alight about where I'm going with this, but I'll take it slow in case you're still confused. Use .split to get the various cleavings of each title. This will be a list of lists.

In [9]:
# examples!
print('chameleon'.split('a')) # one a begets two fragments
print('bargain'.split('a')) # two a begets three fragments
print('avalanche'.split('a')) # three a begets four fragments (note that the .split includes a separate fragment even though a is the first letter!)

# in a list:
example_words = make_array('chameleon', 'bargain', 'avalanche') # this represents the array of titles
split_list = [x.split('a') for x in example_words] # this is a list of lists
split_list

['ch', 'meleon']
['b', 'rg', 'in']
['', 'v', 'l', 'nche']


[['ch', 'meleon'], ['b', 'rg', 'in'], ['', 'v', 'l', 'nche']]

In [10]:
# Use the examples to get the various fragments of the titles in your title column array
split_titles = [x.split('a') for x in lowercase_titles]

Step 3: Use another function in Exercises 02 to get the length of each sub-list in the list. Adjust the values so that they match the number of a's. (Remember, a list of fragments of length 3 came from a title that had TWO a's, not three!)

In [11]:
# example! A function on all sublists of a list. This function is NOT what you want for step 3, however.
[x[0] for x in split_list] # this takes the first term of all the sublists.

['ch', 'b', '']

In [12]:
num_a = [len(x)-1 for x in split_titles]

Step 4: Append the column to the imdb table.

In [13]:
imdb = imdb.with_column("Number of A's", num_a)

What are the five movies with the most A's in the title?

In [14]:
imdb.sort("Number of A's", descending=True)

Votes,Rating,Title,Year,Decade,Title Length,Number of A's
32385,8,La battaglia di Algeri,1966,1960,22,5
466111,8,Harry Potter and the Deathly Hallows: Part 2,2011,2010,44,5
718645,8,Pirates of the Caribbean: The Curse of the Black Pearl,2003,2000,54,5
59377,8,Lagaan: Once Upon a Time in India,2001,2000,33,5
453748,8.3,Indiana Jones and the Last Crusade,1989,1980,34,5
170302,8.3,Lawrence of Arabia,1962,1960,18,4
527349,8,Guardians of the Galaxy,2014,2010,23,4
350551,8.5,Casablanca,1942,1940,10,4
81809,8.1,Kaze no tani no Naushika,1984,1980,24,4
46987,8,Relatos salvajes,2014,2010,16,3


The Battle of Algiers, HPDH2, PotC Black Pearl, Lagaan, and IJ lost crusade

### 1.3: Date Formatting
Now time for some actually-real-world applications. A very common sight in Tables and Data Statistics in general is the reformatting of dates to render them more code-oriented. After all, Python has no idea what the slashes in '7/31/2025' mean, nor does it comprehend the colons in '6:30:25'. To be able to work with times and dates, we need to reexpress these values in terms of a single unit (e.g. hours) instead of multiple units (e.g. hours : minutes : seconds).

Consider the times in the following table. Our task in this exercise is to convert them from hh:mm:ss to some amount of hours past midnight. For example, a time of 11:30:00 gets a value of 11.5 hours past midnight, and 7:25:52 gets a value of 7.431 hours past midnight. In essence, we are doing the following: (using 15:35:20 as an example)
1. Getting the seconds in terms of a decimal portion of a minute (e.g. 20 seconds is 1/3 of a minute or 0.33 minutes)
2. Adding that to the minute count (35+0.33 = 35.33 minutes)
3. Expressing that minute count in terms of a decimal portion of an hour (35.33 minutes becomes 0.589 hours)
4. Adding that to the hours (15+0.589 = 15.589 hours)

In [15]:
unformatted = ["4:50:31", "7:12:45", "9:38:09", "11:27:56", "13:03:22", "15:41:10", "17:19:44", "20:02:37", "22:46:05", "00:15:28"]

times = Table().with_columns('hh:mm:ss times', unformatted)
times # this is typically what these times will look like within a table. 
# Of course, other data will also be there, but for the sake of simplicity, this is not included.

hh:mm:ss times
4:50:31
7:12:45
9:38:09
11:27:56
13:03:22
15:41:10
17:19:44
20:02:37
22:46:05
00:15:28


When having to format a whole list, it's a good idea to focus solely on one singular value, and then extrapolate to the entire thing. Try working with just this first value below.

In [16]:
test_time = times[0][0]
test_time

'4:50:31'

In [17]:
components = test_time.split(':')
float_components = [float(x) for x in components]
float_components

[4.0, 50.0, 31.0]

In [18]:
M_and_S = float_components[1]+float_components[2]/60

In [19]:
Total_hours = float_components[0] + M_and_S/60
Total_hours

4.841944444444445

Here are some guiding questions:
* How do you 'unpack' the values in the data format?
* Are these unpacked values able to be worked with numerically? What do you have to do to make them so?
* Once you have have worked it out and gotten the singular test_time to work, what do you have to do to apply this to the whole table column?
* Remember, what first step do you have to do to make data within tables open to analysis?

Once you've solved it, append the formatted times as another column to the table.

In [20]:
raw = times.column('hh:mm:ss times')
raw_split = [x.split(':') for x in raw]
float_split = [[float(y) for y in x] for x in raw_split] # ok this may have been an oversight on how difficult it is
float_split

[[4.0, 50.0, 31.0],
 [7.0, 12.0, 45.0],
 [9.0, 38.0, 9.0],
 [11.0, 27.0, 56.0],
 [13.0, 3.0, 22.0],
 [15.0, 41.0, 10.0],
 [17.0, 19.0, 44.0],
 [20.0, 2.0, 37.0],
 [22.0, 46.0, 5.0],
 [0.0, 15.0, 28.0]]

In [21]:
col_hours = [x[0]+(x[1]+x[2]/60)/60 for x in float_split]
col_hours

[4.841944444444445,
 7.2125,
 9.635833333333334,
 11.465555555555556,
 13.056111111111111,
 15.686111111111112,
 17.32888888888889,
 20.04361111111111,
 22.768055555555556,
 0.2577777777777778]

In [22]:
times = times.with_column('hours past midnight',col_hours)
times # i didn't expect 1.3 to be this hard. I might drop it in the future.

hh:mm:ss times,hours past midnight
4:50:31,4.84194
7:12:45,7.2125
9:38:09,9.63583
11:27:56,11.4656
13:03:22,13.0561
15:41:10,15.6861
17:19:44,17.3289
20:02:37,20.0436
22:46:05,22.7681
00:15:28,0.257778


## Exercise 2:
A crucial part of data science is being able to translate abstract questions into code lines that will give you the answer to the question. This type of translation is akin to mathematical calculations or finding a route to drive to a certain place. You've already seen such questions in the lab itself, but given how important they are to your projects, you will need plenty more practice for this. The main strategy is to know what your answer should look like and how to get there, and therefore to design a series of steps to get from the entire table to your final destination.

### 2.1: World Populations
These questions regard the dataset 'world_populations.csv'. I've done some prep work below to include the years.

In [23]:
pops = Table.read_table('world_population.csv')
pops = pops.with_column('Year', np.arange(1950, 2016))
pops

Population,Year
2557628654,1950
2594939877,1951
2636772306,1952
2682053389,1953
2730228104,1954
2782098943,1955
2835299673,1956
2891349717,1957
2948137248,1958
3000716593,1959


How many years did the world spend with a population between 4 and 5 billion?

In [24]:
pops.where('Population', are.between(4000000000,5000000000)).num_rows


13

Has the world population grown a larger amount from 1950 to 2000 or from 2000 to 2016? By how much are these amounts different?

In [25]:
fifties_to_oughts = pops.where('Year', are.between(1950,2001))
oughts_to_sixteen = pops.where('Year', are.between(2000,2017)) 
fif_to_ought_diff = fifties_to_oughts.column('Population')[-1]-fifties_to_oughts.column('Population')[0]
ought_to_sixteen_diff = oughts_to_sixteen.column('Population')[-1]-oughts_to_sixteen.column('Population')[0]
fif_to_ought_diff-ought_to_sixteen_diff # world has grown about 2.36 billion more from 1950 to 2000 than from 2000 to 2016. 

2363024101

### 2.2: Farmers Markets
These questions regard the dataset 'farmers_markets.csv'.

In [26]:
farmers_markets = Table().read_table('farmers_markets.csv')
farmers_markets = farmers_markets.drop('FMID','Website','Facebook','Twitter','Youtube','OtherMedia','Season1Date','Season1Time','Season2Date','Season2Time','Season3Date','Season3Time','Season4Date','Season4Time')
farmers_markets

MarketName,street,city,County,State,zip,x,y,Location,Credit,WIC,WICcash,SFMNP,SNAP,Organic,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,Tofu,WildHarvested,updateTime
Caledonia Farmers Market Association - Danville,nan,Danville,Caledonia,Vermont,05828,-72.1403,44.411,nan,Y,Y,N,Y,N,Y,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,Y,Y,Y,Y,Y,N,Y,Y,Y,N,Y,N,Y,N,N,6/28/2016 12:10:09 PM
Stearns Homestead Farmers' Market,6975 Ridge Road,Parma,Cuyahoga,Ohio,44130,-81.7286,41.3751,nan,Y,Y,N,Y,Y,-,Y,N,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,Y,N,N,N,N,N,N,N,Y,N,N,N,Y,N,N,4/9/2016 8:05:17 PM
100 Mile Market,507 Harrison St,Kalamazoo,Kalamazoo,Michigan,49007,-85.5749,42.296,nan,Y,Y,N,Y,Y,N,Y,Y,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,N,Y,Y,Y,N,Y,N,N,Y,Y,N,N,N,N,N,4/16/2016 12:37:56 PM
106 S. Main Street Farmers Market,106 S. Main Street,Six Mile,nan,South Carolina,29682,-82.8187,34.8042,nan,Y,N,N,N,N,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,2013
10th Steet Community Farmers Market,10th Street and Poplar,Lamar,Barton,Missouri,64759,-94.2746,37.4956,nan,Y,N,N,N,N,-,Y,N,Y,N,Y,N,Y,Y,Y,Y,N,Y,N,N,Y,Y,Y,Y,N,N,N,N,Y,N,N,N,N,N,N,10/28/2014 9:49:46 AM
112st Madison Avenue,112th Madison Avenue,New York,New York,New York,10029,-73.9493,40.7939,Private business parking lot,N,N,Y,Y,N,-,Y,N,Y,Y,N,N,Y,Y,Y,Y,N,N,N,Y,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,N,3/1/2012 10:38:22 AM
12 South Farmers Market,3000 Granny White Pike,Nashville,Davidson,Tennessee,37204,-86.7907,36.1184,nan,Y,N,N,N,Y,Y,Y,Y,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,N,Y,Y,Y,N,N,Y,N,Y,N,Y,Y,Y,N,N,5/1/2015 10:40:56 AM
125th Street Fresh Connect Farmers' Market,"163 West 125th Street and Adam Clayton Powell, Jr. Blvd.",New York,New York,New York,10027,-73.9482,40.809,Federal/State government building grounds,Y,Y,N,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,Y,N,Y,Y,Y,N,Y,Y,N,Y,N,Y,N,N,N,N,4/7/2014 4:32:01 PM
12th & Brandywine Urban Farm Market,12th & Brandywine Streets,Wilmington,New Castle,Delaware,19801,-75.5345,39.7421,"On a farm from: a barn, a greenhouse, a tent, a stand, etc",N,N,N,N,Y,N,N,N,N,N,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,4/3/2014 3:43:31 PM
14&U Farmers' Market,1400 U Street NW,Washington,District of Columbia,District of Columbia,20009,-77.0321,38.917,Other,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,Y,N,Y,Y,Y,Y,N,Y,N,Y,Y,Y,N,N,N,N,N,Y,Y,Y,Y,N,N,N,N,4/5/2014 1:49:04 PM


What percentage of these markets have 'Farmers' in their name?

In [27]:
farmer_farmer = farmers_markets.where('MarketName', are.containing('Farmers'))
100*farmer_farmer.num_rows/farmers_markets.num_rows

75.02925345190732

What markets lie within 1 degree, latitude and longitude, of the Temple University's SERC building (the coordinates are -75.153, 39.982)? For reference, one degree of latitude/longitude is equal to 69 miles. (nice)

In [28]:
farmers_markets.where('x', are.between(-76.153, -74.153)).where('y', are.between(38.982, 40.982)) 
# they can just show the table; don't want them to list off ~250 markets.

MarketName,street,city,County,State,zip,x,y,Location,Credit,WIC,WICcash,SFMNP,SNAP,Organic,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,Tofu,WildHarvested,updateTime
12th & Brandywine Urban Farm Market,12th & Brandywine Streets,Wilmington,New Castle,Delaware,19801,-75.5345,39.7421,"On a farm from: a barn, a greenhouse, a tent, a stand, etc",N,N,N,N,Y,N,N,N,N,N,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,4/3/2014 3:43:31 PM
18th and Christian Farmers' Market,18th Street and Christian Street,Philadelphia,Philadelphia,Pennsylvania,19146,-75.1734,39.9409,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 3:33:48 PM
22nd and Tasker Farmers' Market,22nd and Tasker Streets,Philadelphia,Philadelphia,Pennsylvania,19146,-75.1813,39.9323,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:29:42 PM
26th and Allegheny,26th Street and W Allegheny Avenue,Philadelphia,Philadelphia,Pennsylvania,19129,-75.1723,40.0041,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 3:17:20 PM
29th and Wharton Farmers' Market,29th and Wharton Streets,Philadelphia,Philadelphia,Pennsylvania,19146,-75.192,39.9373,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:47:10 PM
33rd and Diamond Farmers' Market,N 33rd and Diamond Streets,Philadelphia,Philadelphia,Pennsylvania,19121,-75.1872,39.9888,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:28:01 PM
4th and Lehigh Farmers' Market,N 4th Street and W. Lehigh Avenue,Philadelphia,Philadelphia,Pennsylvania,19133,-75.1387,39.9917,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 3:11:40 PM
52nd and Haverford Farmers' Market,N 52nd Street and Haverford Avenue,Philadelphia,Philadelphia,Pennsylvania,19139,-75.2245,39.9651,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:38:14 PM
58th and Chester Farmers' Market,58th Street and Chester Avenue,Philadelphia,Philadelphia,Pennsylvania,19143,-75.2283,39.9357,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:35:04 PM
ABV Farm Market,690 Allegheny Road,Mount Bethe,nan,Pennsylvania,18343,-75.0998,40.8743,nan,Y,N,N,N,N,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,2013


What markets sell Meat and Tofu but no vegetables?

In [29]:
farmers_markets.where('Meat', 'Y').where('Tofu', 'Y').where('Vegetables', 'N')

MarketName,street,city,County,State,zip,x,y,Location,Credit,WIC,WICcash,SFMNP,SNAP,Organic,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,Tofu,WildHarvested,updateTime
Old Town Alexandria Farmers' Market,301 King Street,Alexandria,Alexandria,Virginia,22314,-77.043,38.8048,nan,Y,N,N,N,Y,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,Y,N,Y,N,Y,Y,Y,Y,Y,Y,N,Y,N,Y,N,Y,Y,N,Y,N,7/11/2015 6:34:15 AM
Saratoga Certified Farmers' Market,"West Valley College, 14000 Fruitvale Avenue",Saratoga,Santa Clara,California,95070,-122.011,37.2642,Federal/State government building grounds,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,N,Y,Y,N,Y,N,Y,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,Y,N,4/1/2014 4:43:25 PM
Valley Farmers' Market Association LLC,"3201 ""F"" Street",Bakersfield,Kern,California,nan,-119.025,35.3886,Private business parking lot,Y,Y,N,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,N,Y,Y,N,Y,N,Y,Y,N,Y,Y,Y,N,Y,N,Y,N,Y,N,Y,Y,N,4/1/2014 9:41:05 AM


## Exercise 3:
### 3.1:
In this section, we will introduce to ourselves a very useful function whose name I will not spoil yet. This involves the use of the farmers_markets table. If you look at the bunch of columns at the far right end, most of them show whether or not a given market sells a particular good. A 'Y' means that the market does sell it, and 'N' means that it doesn't.

Suppose you pick the Cheese column (because I like cheese) that has these Ys and Ns and want to know how many of them have Ys (how many sell cheese) and how many of them have Ns (how many don't sell cheese). (Note that Ys and Ns are the only inputs in the column.) How do you do this?

Step 1: Define the `Y_cheese` variable as the farmers_markets table containing only the rows with 'Y' in the Cheese column, and the `N_cheese` variable as the table containing only the rows with 'N' in the Cheese column.

In [30]:
Y_cheese = farmers_markets.where('Cheese', 'Y')
N_cheese = farmers_markets.where('Cheese', 'N')

Step 2: Get the number of rows in the Y_cheese and N_cheese, and define them as `Y_rows` and `N_rows`.

In [31]:
Y_rows = Y_cheese.num_rows
N_rows = N_cheese.num_rows

Note that what you've gotten right now, in Y_rows and N_rows, is the number of markets that have cheese (number of rows that have 'Y') and don't have cheese (number of rows that have 'N'). Now, assuming you've done your duty in properly defining these variables, I'll package this information in a format that is much more pleasant to your eyes.

In [32]:
# don't change this cell!
inputs = ['Y', 'N']
number_rows = [Y_rows, N_rows]
summary_cheese = Table().with_columns('Cheese', inputs, 'count', number_rows)
summary_cheese = summary_cheese.sort('Cheese')
summary_cheese

Cheese,count
N,5802
Y,2744


This is the essence of our mystery function: for each unique element in a given column, it will tell you how many rows in the table have that input for that column. Here is that mystery function below.

In [33]:
farmers_markets.group('Cheese')

Cheese,count
N,5802
Y,2744


### 3.2: Practice with `.group`
Which state has the most markets?

In [34]:
states = farmers_markets.group('State')
states.sort('count', descending=True)

State,count
California,755
New York,672
Michigan,340
Ohio,326
Illinois,323
Massachusetts,316
Wisconsin,307
Pennsylvania,294
Missouri,258
Virginia,253


Going back to the imdb table, which decade had the most movies?

In [35]:
decades=imdb.group('Decade')
decades.sort('count', descending=True)

Decade,count
2000,50
1990,42
1980,31
1950,30
2010,29
1960,22
1970,21
1940,14
1930,7
1920,4


What's the maximum number of A's a movie had? Create the .group table and assign it the variable `A_stats`.

Note that, since the .group method gives you a table, you can modify that table in itself. Use this fact to add on to `A_stats` another column that tells you what percentage of the total movies had that number of A's. (e.g. What percent had zero A's? What percent had one A? and so forth)

In [36]:
A_stats = imdb.group("Number of A's")
A_stats

Number of A's,count
0,81
1,84
2,51
3,25
4,4
5,5


In [37]:
A_count = A_stats.column('count')
A_percentage = [x/sum(A_count) for x in A_count]
A_stats = A_stats.with_column('percentage', A_percentage)
A_stats

Number of A's,count,percentage
0,81,0.324
1,84,0.336
2,51,0.204
3,25,0.1
4,4,0.016
5,5,0.02


### 3.3: `.group` with a function
When paired with a function as the second argument, the resultant table is very different. Let's look back at the imdb table, and apply the .group method and the np.average function to the 'Decade' column. It does filter the table according to each unique input of Decade, but instead of taking the number of rows for each input in Decade, it looks at the other columns besides Decade and takes their average (applies the specified function).

In [38]:
imdb.group('Decade', np.average)

Decade,Votes average,Rating average,Title average,Year average,Title Length average,Number of A's average
1920,64854,8.25,,1924.75,10.25,0.25
1930,125825,8.27143,,1935.57,15.2857,0.428571
1940,122767,8.25714,,1944.07,16.7143,1.35714
1950,109829,8.23333,,1955,15.0667,0.833333
1960,147391,8.23182,,1964.14,21.1364,1.81818
1970,353377,8.30952,,1975.62,12.8571,1.09524
1980,295290,8.24194,,1984.29,16.1935,1.74194
1990,559449,8.35714,,1995.19,14.5238,0.952381
2000,481929,8.26,,2004.26,18.44,1.14
2010,438006,8.2,,2012.31,16.1034,1.37931


Let's take the first row as an example. 

What it tells us is that, for all the 1920 movies in the table...
* They got 64854 votes on average
* They had an average rating of 8.25
* The average year in which they were released was 1924 (ignoring decimals)
* They had an average of 0.25 A's in their name.

And this goes for all the movies released in the decade of 1920. The 1930 movies' average stats are in the second row, the 40s' stats in the third, and so forth. 

Note that the Title column is blank because you can't take the average of non-numerical values.

**What functions can you use in `.group`?**

Any function works SO LONG AS its only input is a list or array. Methods do not work, as the .group assumes that this only input is in the parentheses following the function name, and the method's input is before it.

**Determine if the following questions can be answered solely through a single instance of `.group`. If yes, answer them through code. If not, formulate a plan to answer them (through establishing steps like I do in Exercise 1.2), but you don't have to actually code it.**

Assume that you have already done 1.2 and thus have the 'Number of As' column at your disposal.

What's the total number of A's in movies of the 1940s?

In [39]:
# Yep, you can do that with a single .group
imdb.group('Decade', sum)

Decade,Votes sum,Rating sum,Title sum,Year sum,Title Length sum,Number of A's sum
1920,259416,33,,7699,41,1
1930,880776,57.9,,13549,107,3
1940,1718735,115.6,,27217,234,19
1950,3294871,247,,58650,452,25
1960,3242603,181.1,,43211,465,40
1970,7420922,174.5,,41488,270,23
1980,9153992,255.5,,61513,502,54
1990,23496861,351,,83798,610,40
2000,24096443,413,,100213,922,57
2010,12702167,237.8,,58357,467,40


For each state in `farmers_markets`, what's the city with the most markets?

In [43]:
# Nope, this requires a bit more than a single .group (in fact, this requires a full-on for-in loop, which I will do here for posterity
max_cities = []
frequency = []
for i in farmers_markets.group('State').column(0):
    cities = farmers_markets.where('State', i).select('city')
    city_count = cities.group('city').sort('count', descending=True)
    biggest_city = city_count[0][0]
    biggest_count = city_count[1][0]
    max_cities.append(biggest_city)
    frequency.append(biggest_count)
State_cities = Table().with_columns('State', farmers_markets.group('State').column(0), 'Most Markets City (MMC)', max_cities, 'Count in MMC', frequency)
State_cities

State,Most Markets City (MMC),Count in MMC
Alabama,Birmingham,5
Alaska,Anchorage,11
Arizona,Tucson,15
Arkansas,Little Rock,7
California,Los Angeles,28
Colorado,Denver,13
Connecticut,Hartford,8
Delaware,Wilmington,10
District of Columbia,Washington,26
Florida,Miami,11


What percentage of markets have a number in their names?

In [45]:
# Nope, this requires more than a singular .group
count = 0
for i in np.arange(0,10):
    markets_with_number = farmers_markets.where('MarketName', are.containing(str(i))).num_rows
    count = count+markets_with_number
100*count/farmers_markets.num_rows

2.375380294874795

What decade had the longest-titled movie names?

In [52]:
# one could reasonably interpret this any one of the following methods. Both of them lead to the same conclusion (!) but I show both here.
# This one uses max() to get the absolute longest title length
imdb.group('Decade', max).sort('Title Length max', descending=True) 

Decade,Votes max,Rating max,Title max,Year max,Title Length max,Number of A's max
1960,447875,8.9,Yôjinbô,1969,68,5
2000,1473049,8.9,Yip Man,2009,54,5
1980,700283,8.7,Tonari no Totoro,1989,46,5
2010,1271949,8.7,X-Men: Days of Future Past,2015,44,5
1990,1498733,9.2,Unforgiven,1999,42,3
1940,350551,8.6,The Treasure of the Sierra Madre,1949,32,4
1970,1027398,9.2,The Sting,1979,31,2
1930,259235,8.5,The Wizard of Oz,1939,28,1
1950,384187,8.9,Witness for the Prosecution,1959,28,3
1920,98794,8.3,The Kid,1927,13,1


In [53]:
# this one uses np.average() to get the decade that consistency had the longest movie names.
imdb.group('Decade', np.average).sort('Title Length average', descending=True)

Decade,Votes average,Rating average,Title average,Year average,Title Length average,Number of A's average
1960,147391,8.23182,,1964.14,21.1364,1.81818
2000,481929,8.26,,2004.26,18.44,1.14
1940,122767,8.25714,,1944.07,16.7143,1.35714
1980,295290,8.24194,,1984.29,16.1935,1.74194
2010,438006,8.2,,2012.31,16.1034,1.37931
1930,125825,8.27143,,1935.57,15.2857,0.428571
1950,109829,8.23333,,1955,15.0667,0.833333
1990,559449,8.35714,,1995.19,14.5238,0.952381
1970,353377,8.30952,,1975.62,12.8571,1.09524
1920,64854,8.25,,1924.75,10.25,0.25


What's the westernmost market that sells Eggs?

In [60]:
# if you wanted JUST to get the latitude (the east-west location), it would be solvable through .group (and many other methods)
farmers_markets.group('Eggs', min)

Eggs,MarketName min,street min,city min,County min,State min,zip min,x min,y min,Location min,Credit min,WIC min,WICcash min,SFMNP min,SNAP min,Organic min,Bakedgoods min,Cheese min,Crafts min,Flowers min,Seafood min,Herbs min,Vegetables min,Honey min,Jams min,Maple min,Meat min,Nursery min,Nuts min,Plants min,Poultry min,Prepared min,Soap min,Trees min,Wine min,Coffee min,Beans min,Fruits min,Grains min,Juices min,Mushrooms min,PetFood min,Tofu min,WildHarvested min,updateTime min
N,106 S. Main Street Farmers Market,"E. 3rd St.,",Dayton,ALAMEDA,Alabama,00000,-166.54,17.7099,Closed-off public street,N,N,N,N,N,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,1/14/2012 4:29:04 PM
Y,Caledonia Farmers Market Association - Danville,1102 McConnell Road,New Canaan,Abbeville,Alabama,0033,-159.396,18.0093,Closed-off public street,N,N,N,N,N,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,1/1/2015 10:12:55 PM


In [59]:
# instead, the question wants the market itself. This actually doesn't require .group.
yeggs = farmers_markets.where('Eggs', 'Y')
yeggs.sort('x')['MarketName'][0]

'Kauai Community Market'